File: src/1124.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int longestWPI(int* hours, int hoursSize) {
    int retVal = 0;

    int i;

    int prefixSum[hoursSize + 1];
    memset(prefixSum, 0, sizeof(prefixSum));
    int monotonicStack[hoursSize + 1];
    memset(monotonicStack, 0, sizeof(monotonicStack));
    int monotonicStackTop = 0;

    monotonicStack[monotonicStackTop++] = 0;
    prefixSum[0] = 0;
    for (i = 1; i <= hoursSize; ++i) {
        prefixSum[i] = prefixSum[i - 1] + (hours[i - 1] > 8 ? 1 : -1);
        if (prefixSum[monotonicStack[monotonicStackTop - 1]] > prefixSum[i]) {
            monotonicStack[monotonicStackTop++] = i;
        }
    }

    for (i = hoursSize; i >= 1; --i) {
        while ((monotonicStackTop != 0) && (prefixSum[monotonicStack[monotonicStackTop - 1]] < prefixSum[i])) {
            retVal = fmax(retVal, i - monotonicStack[monotonicStackTop - 1]);
            --monotonicStackTop;
        }
    }

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (int)(1e4)
    struct testCaseType {
        int hours[MAX_SIZE];
        int hoursSize;
    } testCase[] = {
        {{9, 9, 6, 0, 6, 6, 9}, 7}, {{6, 6, 6}, 3}, {{9, 9, 6, 9, 0, 6, 9, 6, 9}, 9}, {{6, 9, 6}, 3}, {{6, 9, 9}, 3}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: hours = [9,9,6,0,6,6,9]
     *  Output: 3
     *
     *  Input: hours = [6,6,6]
     *  Output: 0
     *
     *  Input: hours = [9,9,6,9,0,6,9,6,9]
     *  Output: 9
     *
     *  Input: hours = [6,9,6]
     *  Output: 1
     *
     *  Input: hours = [6,9,9]
     *  Output: 3
     */

    int answer = 0;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: hours = [");
        for (j = 0; j < testCase[i].hoursSize; ++j) {
            printf("%s%d", (j == 0 ? "" : ","), testCase[i].hours[j]);
        }
        printf("]\n");

        answer = longestWPI(testCase[i].hours, testCase[i].hoursSize);
        printf("Output: %d\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}

File: src/1124.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def longestWPI(self, hours: List[int]) -> int:
        retVal = 0

        hoursSize = len(hours)

        prefixSum = [0] * (hoursSize + 1)
        monotonicStack = [0]
        for idx, hour in enumerate(hours, 1):
            prefixSum[idx] = prefixSum[idx - 1] + (1 if hour > 8 else -1)
            if prefixSum[idx] < prefixSum[monotonicStack[-1]]:
                monotonicStack.append(idx)

        for i in range(hoursSize, 0, -1):
            while monotonicStack and prefixSum[i] > prefixSum[monotonicStack[-1]]:
                retVal = max(retVal, i - monotonicStack.pop())

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for hours in [[9, 9, 6, 0, 6, 6, 9], [6, 6, 6], [9, 9, 6, 9, 0, 6, 9, 6, 9], [6, 9, 6], [6, 9, 9]]:
            # /* Example
            #  *  Input: hours = [9,9,6,0,6,6,9]
            #  *  Output: 3
            #  *
            #  *  Input: hours = [6,6,6]
            #  *  Output: 0
            #  *
            #  *  Input: hours = [9,9,6,9,0,6,9,6,9]
            #  *  Output: 9
            #  *
            #  *  Input: hours = [6,9,6]
            #  *  Output: 1
            #  *
            #  *  Input: hours = [6,9,9]
            #  *  Output: 3
            #  */
            logging.info("Input: hours = %s", hours)

            retVal = pSolution.longestWPI(hours)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)
